### PR TITLE
feature: added backend client, refactored to async

### PR DESCRIPTION
**Pull request by 1ex**

### CODE REVIEW

File: neurons/validator/services/openai/service.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+                top_p=1,
+                frequency_penalty=0,
+                presence_penalty=0,
+            )
+        except Exception as e:
+            logger.error(f"openai completion request failed: {e}")
+            raise OpenAIRequestFailed(str(e)) from e
+
+        logger.info(f"OpenAI response object: {response}")
+        response = response.choices[0].message.content
```
Comment (tavurth) [inline neurons/validator/services/openai/service.py]: Can fail, choices can be `[]`

File: neurons/validator/services/openai/service.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        Returns True if prompt contains any nsfw content
+        """
+        try:
+            response = await self.openai_client.moderations.create(input=prompt)
+        except Exception as e:
+            logger.error(f"[check_prompt_for_nsfw] failed to do openai request: {e}")
+            raise OpenAIRequestFailed(str(e)) from e
+
+        # Check if the moderation flagged the prompt as NSFW
+        moderation_results = response.results[0]
```
Comment (tavurth) [inline neurons/validator/services/openai/service.py:81]: Also can fail `results = []`

File: neurons/validator/validator.py [hunk 3]
```diff
@@ -415,7 +359,61 @@ async def run(self):
...
+        # before going on and creating a synthetic task
+
+        logger.info(
+            f"polling backend for incoming image generation task ({timeout}s) ..."
+        )
+
+        task = await get_task(self.backend_client, timeout=timeout)
+
+        # No organic task found
+        if task is None and (prompt := await generate_random_prompt_gpt(self)):
```
Comment (tavurth) [inline neurons/validator/validator.py]: Please move to a newline, this is similar to `ternary in conditional` makes things hard to read

File: neurons/validator/utils.py [hunk 4]
```diff
@@ -43,33 +50,26 @@ def get_validator_spec_version() -> int:
...
+    )
+    async def _get_task_with_timeout():
+        try:
+            return await backend_client.get_task(timeout=1)
+        except GetTaskError as e:
+            logger.error(f"error getting task: {e}")
+            return None
 
-    return task
+    return await _get_task_with_timeout()
```
Comment (tavurth) [inline neurons/validator/utils.py]: Could this not be moved into backend client?

Comment (1ex) [reply neurons/validator/utils.py]: actually this is *polling* logic, so i kept this outside of client, i will move this method to backend yeah

File: neurons/validator/validator.py (hunk 3 above)
Comment (tavurth) [reply neurons/validator/validator.py]: Aha, it actually leads to a bug, so yea we should certainly not do this type.

`task is None` but `gpt` fails to generate, which leads to returning `None` from the function @1ex

However `None` is also checked for NSFW and `task.prompt` raises an `Exception`

**Pull request merged**